File: src/main/java/com/manabreakstudios/discourse/ui/core/editor/Grid.java
package com.manabreakstudios.discourse.ui.core.editor;

import com.manabreakstudios.discourse.ui.Theme;
import lombok.Getter;

import java.awt.*;

public class Grid {

    @Getter
    private int size = 30;

    public Point snap(Point point) {
        return new Point((int) (point.getX() / (float) size) * size, (int) (point.getY() / (float) size) * size);
    }

    public void paint(Graphics g, int x, int y, int w, int h) {
        int firstVisibleLine = 0;
        int firstVisibleColumn = 0;

        Graphics2D g2d = (Graphics2D) g.create();

        g2d.setColor(Theme.theme().getGridColor());
        for (int l = 0; firstVisibleLine + l * size < h; l += 1) {
            int lineY = firstVisibleLine + l * size;
            g2d.drawLine(0, lineY, w, lineY);
        }
        for (int c = 0; firstVisibleColumn + c * size < w; c += 1) {
            int lineX= firstVisibleColumn + c * size;
            g2d.drawLine(lineX, 0, lineX, h);
        }

        g2d.setColor(Theme.theme().getGridMajorColor());
        for (int l = 0; firstVisibleLine + l * size < h; l += 4) {
            int lineY = firstVisibleLine + l * size;
            g2d.drawLine(0, lineY, w, lineY);
        }
        for (int c = 0; firstVisibleColumn + c * size < w; c += 4) {
            int lineX= firstVisibleColumn + c * size;
            g2d.drawLine(lineX, 0, lineX, h);
        }

        g2d.dispose();
    }
}

File: node-editor/src/com/manabreak/node_editor/Grid.kt
package com.manabreak.node_editor

import com.manabreak.node_editor.Theme.Companion.theme
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.Point

class Grid {

    val size = 30

    fun snap(point: Point): Point {
        return Point((point.getX() / size.toFloat()).toInt() * size, (point.getY() / size.toFloat()).toInt() * size)
    }

    fun paint(g: Graphics, x: Int, y: Int, w: Int, h: Int) {
        val firstVisibleLine = 0
        val firstVisibleColumn = 0

        val g2d = g.create() as Graphics2D

        g2d.color = theme.gridColor
        run {
            var l = 0
            while (firstVisibleLine + l * size < h) {
                val lineY = firstVisibleLine + l * size
                g2d.drawLine(0, lineY, w, lineY)
                l += 1
            }
        }
        run {
            var c = 0
            while (firstVisibleColumn + c * size < w) {
                val lineX = firstVisibleColumn + c * size
                g2d.drawLine(lineX, 0, lineX, h)
                c += 1
            }
        }

        g2d.color = theme.gridMajorColor
        var l = 0
        while (firstVisibleLine + l * size < h) {
            val lineY = firstVisibleLine + l * size
            g2d.drawLine(0, lineY, w, lineY)
            l += 4
        }
        var c = 0
        while (firstVisibleColumn + c * size < w) {
            val lineX = firstVisibleColumn + c * size
            g2d.drawLine(lineX, 0, lineX, h)
            c += 4
        }

        g2d.dispose()
    }
}
